stan/math/prim/err/elementwise_check.hpp

```diff
@@ -201,21 +99,119 @@ class Iser {
  * `x` is a container, check each element inside `x`, recursively.
  * @tparam F type of predicate
  * @tparam T type of `x`
+ * @tparam Indexings types of `indexings`
  * @param is_good predicate to check, must accept doubles and produce bools
  * @param function function name (for error messages)
  * @param name variable name (for error messages)
  * @param x variable to check, can be a scalar, a container of scalars, a
  * container of containers of scalars, etc
- * @param suffix message to print at end of error message
+ * @param must_be message describing what the value should be
+ * @param indexings any additional indexing to print. Intended for internal use
+ * in `elementwise_check` only.
  * @throws `std::domain_error` if `is_good` returns `false` for the value
  * of any element in `x`
  */
-template <typename F, typename T>
+template <typename F, typename T, typename... Indexings,
+          require_stan_scalar_t<T>* = nullptr>
+inline void elementwise_check(const F& is_good, const char* function,
+                              const char* name, const T& x, const char* must_be,
+                              const Indexings&... indexings) {
+  if (unlikely(!is_good(value_of_rec(x)))) {
+    internal::elementwise_throw_domain_error(function, ": ", name, indexings...,
+                                             " is ", x, ", but must be ",
+                                             must_be, "!");
+  }
+}
+template <typename F, typename T, typename... Indexings,
+          require_eigen_t<T>* = nullptr,
+          std::enable_if_t<static_cast<bool>(Eigen::internal::traits<T>::Flags&(
+              Eigen::LinearAccessBit | Eigen::DirectAccessBit))>* = nullptr>
+inline void elementwise_check(const F& is_good, const char* function,
+                              const char* name, const T& x, const char* must_be,
+                              const Indexings&... indexings) {
+  for (size_t i = 0; i < x.size(); i++) {
+    auto scal = value_of_rec(x.coeff(i));
+    if (unlikely(!is_good(scal))) {
+      if (is_eigen_vector<T>::value) {
+        internal::elementwise_throw_domain_error(
+            function, ": ", name, indexings..., "[", i + error_index::value,
+            "] is ", scal, ", but must be ", must_be, "!");
+      } else if (Eigen::internal::traits<T>::Flags & Eigen::RowMajorBit) {
+        internal::elementwise_throw_domain_error(
+            function, ": ", name, indexings..., "[",
+            i / x.rows() + error_index::value, ", ",
+            i % x.rows() + error_index::value, "] is ", scal, ", but must be ",
```

Comment:
If it's row major, I think this should be `/ x.cols()` and `% x.cols()` (cause `x.cols()` is the length of a row)

Reply:
Fixed and improved tests to catch that.